Automattic/jetpack
# Gutenberg: Register VideoPress extension

**Don't merge**

This includes changes from #11162 that need to be merged first.

### Changes proposed in this Pull Request:


- [x] Add a new `jetpack_register_extension` utility for registering generic extensions that are neither blocks nor plugins. The registration is handled by the callback function passed as argument.
- [x] Register a new extension called `videopress` using the previous utility. This extension customizes the Core video block content by replacing the video source with the VideoPress URL which links to a video file that is supported in all the browsers.
Since we're actually converting the Core video block to a [dynamic block](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks/), any existing post containing a Core video block will automatically use the VideoPress URL without re-publishing the post. This also makes that videos currently being converted by VideoPress are handled, since we use the original source while the videos are not transcoded.
- [x] Create a new file `gutenberg-extension.php` responsible of the extension registration that will be synced in WordPress.com with Fusion.
- [x] Add VideoPress data to the `media` endpoint so it can be consumed by the client-side. (Moved to an independent PR: https://github.com/Automattic/jetpack/pull/11162).

Part of Automattic/jetpack#8764 and Automattic/wp-calypso#29328.

### Testing instructions:





* Spin up a new Jurassic Ninja test site by visiting https://jurassic.ninja/create?gutenpack&calypsobranch=add/videopress-hook&branch=add/videopress-gutenberg-extension
* Connect Jetpack to a WP.com site with a premium/business plan.
* Go to Settings → Jetpack Constants in the site’s wp-admin and set `JETPACK_BETA_BLOCKS` to true.
* Go to Jetpack → Settings and activate the VideoPress module (by enabling the "Enable high-speed, ad-free video player" option).
* Add a new post.
* Insert a video block.
* Upload a video file in a format that is not supported by your browser (e.g. .mov in Chrome).
* Publish it immediately.
* Check that a black placeholder is displayed on the published view. Note that this will only happen if VideoPress has not finished yet to convert the video.
<img width="333" alt="screen shot 2019-01-14 at 16 42 41" src="https://user-images.githubusercontent.com/1233880/51122999-6f2e4e00-181b-11e9-9583-05f3cb7583bb.png">

* Wait until VideoPress finishes to convert the video (you can check this on the Media section).
* Reload the published view of the post.
* Make sure that the video is now playable.
<img width="547" alt="screen shot 2019-01-14 at 16 48 22" src="https://user-images.githubusercontent.com/1233880/51123371-3b075d00-181c-11e9-900a-99940599b4ad.png">

* Inspect the video player with the browser devtools and confirm that the `src` attribute is pointing to the VideoPress video (`videos.files.wp.com`).
* Repeat the process downgrading the plan / disabling the VideoPress module / disabling the beta extensions and check that video is not previewed on the editor, not playable on the published view and the `src` attribute doesn't point to `videos.files.wp.com`. _Note that you need to upload a new video since the previous one is already converted)._

### Proposed changelog entry for your changes:


* Gutenberg: Register VideoPress extension
